go/parquet/internal/utils/bit_reader_test.go

```diff
@@ -0,0 +1,662 @@
+	assert.False(t, ok)
+}
+
+func TestBitWidthZeroLiteral(t *testing.T) {
+	const ngroups = 4
+	buf := []byte{4<<1 | 1}
+	dec := utils.NewRleDecoder(bytes.NewReader(buf), 0)
+	const nvals = ngroups * 8
+	for i := 0; i < nvals; i++ {
+		val, ok := dec.GetValue()
+		assert.True(t, ok)
+		assert.Zero(t, val)
+	}
+	_, ok := dec.GetValue()
+	assert.False(t, ok)
+}
+
+func (r *RLERandomSuite) checkRoundTrip(vals []uint64, width int) bool {
+	const buflen = 64 * 1024
+	buf := make([]byte, buflen)
+	var encoded int
+
+	res := r.Run("encode values", func() {
+		enc := utils.NewRleEncoder(utils.NewWriterAtBuffer(buf), width)
+		for idx, val := range vals {
+			r.Require().Truef(enc.Put(val), "encoding idx: %d", idx)
+		}
+		encoded = enc.Flush()
+	})
+
+	res = res && r.Run("decode individual", func() {
+		dec := utils.NewRleDecoder(bytes.NewReader(buf[:encoded]), width)
+		for idx, val := range vals {
+			out, ok := dec.GetValue()
+			r.True(ok)
+			r.Require().Equalf(out, val, "mismatch idx: %d", idx)
+		}
+	})
+
+	res = res && r.Run("batch decode", func() {
+		dec := utils.NewRleDecoder(bytes.NewReader(buf[:encoded]), width)
+		read := make([]uint64, len(vals))
+		r.Require().Equal(len(vals), dec.GetBatch(read))
+		r.Equal(vals, read)
+	})
+
+	return res
+}
+
+func (r *RLERandomSuite) checkRoundTripSpaced(vals array.Interface, width int) {
+	nvalues := vals.Len()
+	bufsize := utils.MaxBufferSize(width, nvalues)
+
+	buffer := make([]byte, bufsize)
+	encoder := utils.NewRleEncoder(utils.NewWriterAtBuffer(buffer), width)
+
+	switch v := vals.(type) {
+	case *array.Int32:
+		for i := 0; i < v.Len(); i++ {
+			if v.IsValid(i) {
+				r.Require().True(encoder.Put(uint64(v.Value(i))))
+			}
+		}
+	}
+
+	encodedSize := encoder.Flush()
+
+	// verify batch read
+	decoder := utils.NewRleDecoder(bytes.NewReader(buffer[:encodedSize]), width)
+	valuesRead := make([]uint64, nvalues)
+	val, err := decoder.GetBatchSpaced(valuesRead, vals.NullN(), vals.NullBitmapBytes(), int64(vals.Data().Offset()))
+	r.NoError(err)
+	r.EqualValues(nvalues, val)
+
+	switch v := vals.(type) {
+	case *array.Int32:
+		for i := 0; i < nvalues; i++ {
+			if vals.IsValid(i) {
+				r.EqualValues(v.Value(i), valuesRead[i])
+			}
+		}
+	}
+}
+
+func (r *RLERandomSuite) TestRandomSequences() {
+	const niters = 50
+	const ngroups = 1000
+	const maxgroup = 16
+
+	values := make([]uint64, ngroups+maxgroup)
+	seed := rand.Uint64() ^ (rand.Uint64() << 32)
+	gen := rand.New(rand.NewSource(seed))
+
+	for itr := 0; itr < niters; itr++ {
+		parity := false
+		values = values[:0]
+
+		for i := 0; i < ngroups; i++ {
+			groupsize := gen.Intn(19) + 1
+			if groupsize > maxgroup {
+				groupsize = 1
+			}
+
+			v := uint64(0)
+			if parity {
+				v = 1
+			}
+			for j := 0; j < groupsize; j++ {
+				values = append(values, v)
+			}
+			parity = !parity
+		}
+		r.Require().Truef(r.checkRoundTrip(values, bits.Len(uint(len(values)))), "failing seed: %d", seed)
+	}
+}
+
+// func TestBitRleOverflow(t *testing.T) {
```

Comment:
why commented out?

Reply:
so this is an artifact of an earlier impl and a test i copied from the C++ impl.

Given the way I implemented the RLE by taking in an io.WriterAt, it's not the same as the c++ version which always expects a specifically sized buffer with a maximum length that it tracks. It just writes to the `io.WriterAt` and returns whether or not that was successful, and should fail if it can't write. So this test ends up being not worthwhile or relevant in this implementation, I had intended to potentially come back and work this is but never did. So i'll just delete it for now rather than leaving it commented out here.